# Кластеризація. Лабораторна робота №3.

## 1. Імпортуємо необхідні бібліотеки


In [168]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.neighbors import NearestNeighbors

## 2. Читаємо датасет з інформацією про людей, що отримана з Вікіпедії.

In [148]:
# Читаємо датасет
people = pd.read_csv("./people_wiki.csv")
people

URI  \
0                <http://dbpedia.org/resource/Digby_Morrell>   
1               <http://dbpedia.org/resource/Alfred_J._Lewy>   
2                <http://dbpedia.org/resource/Harpdog_Brown>   
3          <http://dbpedia.org/resource/Franz_Rottensteiner>   
4                       <http://dbpedia.org/resource/G-Enka>   
...                                                      ...   
59066               <http://dbpedia.org/resource/Olari_Elts>   
59067           <http://dbpedia.org/resource/Scott_F._Crago>   
59068  <http://dbpedia.org/resource/David_Cass_(footballer)>   
59069              <http://dbpedia.org/resource/Keith_Elias>   
59070             <http://dbpedia.org/resource/Fawaz_Damrah>   

                          name  \
0                Digby Morrell   
1               Alfred J. Lewy   
2                Harpdog Brown   
3          Franz Rottensteiner   
4                       G-Enka   
...                        ...   
59066               Olari Elts   
59067           Scott F. Crago   
59068  David Cass (footballer)   
59069              Keith Elias   
59070             Fawaz Damrah   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               text  
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        

In [149]:
# Знаходимо статтю про конкретну людину Барак Обама
obama = people[people.name == "Barack Obama"].copy()
obama.iloc[0]

URI                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     

## 3. Отримаємо кількість слів для статті Обами


In [150]:
# Векторизація документа за допомогою техніки BAG_OF_WORDS
vectorizer = CountVectorizer()
# Вивчення всіх унікальних слів в тексті (створення словника) та перетворення тексту в числовий вектор, де кожен елемент кількість появ відповідного слова =>
# отримуємо розріджену матрицю частот (зберігає тільки ненульові значення та їх індекси)
features = vectorizer.fit_transform(obama.text)
vocab = vectorizer.get_feature_names_out()
# Загальна кількість разів коли кожне слово зустрілося
dist = np.sum(features.toarray(), axis=0)

In [151]:
# Створюємо таблицю частот слів
obama_word_count_table = pd.DataFrame({'token': vocab, 'count': dist})
# Відсортуємо її за зменшенням кількості слів та виведемо 10 найчастіших слів
obama_word_count_table.sort_values('count', ascending=False).head(10)

token  count
242    the     40
115     in     30
28     and     21
162     of     18
245     to     14
106    his     11
160  obama      9
18     act      8
104     he      7
133    law      6

## 4. Розрахуємо TF_IDF для всієї таблиці (всього корпусу документів)


In [152]:
vectorizer_tf = TfidfVectorizer(stop_words="english") # Автоматично обчислює TF, зменшує вагу слів якщо низький IDF, одразу виключаємо стоп-слова, які не несуть жодної цінності для алгоритму
dtm = vectorizer_tf.fit_transform(people.text) # TF_IDF матриця (розріджена)

In [153]:
dtm.shape # Розмірність матриці

(59071, 548115)

In [154]:
# Отримуємо список всіх слів, які були знайдені у документах
features_tf = vectorizer_tf.get_feature_names_out()

## 5. Вивчимо TF_IDF для статті Обами

In [155]:
obama

URI          name  \
35817  <http://dbpedia.org/resource/Barack_Obama>  Barack Obama   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 text  
35817  barack hussein obama ii brk husen bm born august 4 1961 is the 44th and current president of the united states and the first african american to hold the office born in honolulu hawaii obama is a graduate of columbia university and harvard law school where he served as president of the harvard law review he was a community organizer in chicago before earning his law degree he worked as a civil rights attorney and taught constitutional law at the university of chicago law school from 1992 to 2004 he served three terms representing the 13th district in the illinois senate from 1997 to 2004 running unsuccessfull

In [156]:
# Знаходимо індекс документа про Обаму
obama_index = obama.index[0]
# Створюємо словник для збереження tf_idf значень
word_scores = {}
for i, word in enumerate(features_tf):
    word_scores[word] = dtm[obama_index, i]
# Створюємо таблицю для відображення tf_idf значень
obama_word_scores = pd.DataFrame(word_scores.items(), columns=['word', 'tfidf'])
obama_word_scores.sort_values('tfidf', ascending=False).head(10)

word     tfidf
358378        obama  0.413495
45070           act  0.282170
259085         iraq  0.171970
292368          law  0.163903
138523      control  0.149369
365249      ordered  0.138633
328423     military  0.135368
155872   democratic  0.129792
417604     response  0.124821
258530  involvement  0.124821

## 6. Розрахунок міри подібності між людьми (Обамою, Клінтоном та Бекхемом) на базі TF_IDF за допомогою функції `cosine_similarity`.

In [157]:
clinton = people[people.name == "Bill Clinton"]
beckham = people[people.name == "David Beckham"]

# Косинусна міра подібності вимірює схожість між двома векторами,
# обчислюючи косинус кута між ними
sim_1 = cosine_similarity(dtm[obama.index], dtm[clinton.index])
sim_2 = cosine_similarity(dtm[obama.index], dtm[beckham.index])

print(f"Косинусна подібність між Обамою та Клінтоном: {sim_1[0]}")
print(f"Косинусна подібність між Обамою та Бекхемом: {sim_2[0]}")

Косинусна подібність між Обамою та Клінтоном: [0.18896718]
Косинусна подібність між Обамою та Бекхемом: [0.02556581]


## 7. Використання моделі NNM (Nearest Neighbour Model)

In [158]:
# Створюємо модель пошуку сусідів, орієнтовану на пошук 10 найближчих
knn_model = NearestNeighbors(n_neighbors=10)
# Навчаємо модель на TF_IDF матриці - надаємо простір для пошуку
knn_model.fit(dtm)

NearestNeighbors(n_neighbors=10)

## 8. Шукаємо найближчих людей до Обами

In [159]:
# Шукаємо найближчих до Обами
distances, indices = knn_model.kneighbors(dtm[obama.index], return_distance=True)
# Виводимо таблицю знайдених найближчих людей
nearest_df = pd.DataFrame({'names': people.loc[indices[0]].name, "distance": distances[0]})
nearest_df.sort_values(by="distance").head(10)

names  distance
35817             Barack Obama  0.000000
24478                Joe Biden  1.165145
38376           Samantha Power  1.207369
57108   Hillary Rodham Clinton  1.219640
38714  Eric Stern (politician)  1.222509
46140             Robert Gibbs  1.236178
18827             Henry Waxman  1.243057
44681   Jesse Lee (politician)  1.244667
6796               Eric Holder  1.248296
2412           Joe the Plumber  1.251607

## Домашнє завдання

 1. Вивести перших 5 слів і їх кількість в тексті про людину по варіанту (__#5 Carrie Underwood__) при представленні тексту за допомогою dtm, а також і TF_IDF.
 Вивести текст про задану людину.

In [160]:
# Виводимо текст про задану людину
carrie = people[people.name == "Carrie Underwood"]
carrie.iloc[0].text

'carrie marie underwood born march 10 1983 is an american singer songwriter and actress she rose to fame as the winner of the fourth season of american idol in 2005 underwood has since become one of the most prominent mainstream recording artists in the music industry setting and breaking several records throughout her career her achievements led her to be inducted into the grand ole opry in 2008 and into the oklahoma music hall of fame in 2009 she has won a vast array of awards including six grammy awards sixteen billboard music awards eleven academy of country music awards and eight american music awards along with being nominated for a golden globe award for best original songher debut album some hearts was released in 2005 bolstered by the huge crossover success of the singles jesus take the wheel and before he cheats it went on to become the fastest selling debut country album in nielsen soundscan history the bestselling solo female debut album in country music history and the bes

In [161]:
# Представляємо текст за допомогою dtm (document-term matrix)
vectorizer_count = CountVectorizer()
# Створюємо словник для статті про Керрі та векторизуємо документ (за BAG_OF_WORDS)
features_1 = vectorizer_count.fit_transform(carrie.text)
vocab_1 = vectorizer_count.get_feature_names_out()
# Загальна кількість разів коли кожне слово зустрілося
count = np.sum(features_1.toarray(), axis=0)
# Створюємо табличне представлення dtm
carrie_word_count_table = pd.DataFrame({'token': vocab_1, 'count': count})
carrie_word_count_table.sort_values("count", ascending=False).head(5)

token  count
184      the     36
24       and     16
130       of     15
98        in     12
56   country     11

In [162]:
# Аналізуємо TF_IDF для даної статті про Керрі
# Створюємо словник для збереження tf_idf значень
word_scores_1 = {}
for i, word in enumerate(features_tf):
    word_scores_1[word] = dtm[carrie.index[0], i]
# Створюємо таблицю для відображення tf_idf значень
carrie_word_count_table = pd.DataFrame(word_scores_1.items(), columns=["token", "tf_idf"])
carrie_word_count_table.sort_values("tf_idf", ascending=False).head(5)

token    tf_idf
508279  underwood  0.519588
141830    country  0.314403
194489     female  0.210140
89613   billboard  0.207110
52523       album  0.187929

2. Вивести відстань між двома людьми по варіанту (__#5 Carrie Underwood and Kelly Ayotte__), також текст про другу людину і провести інтерпретацію результату.

In [163]:
# Знаходимо рядок Kelly Ayotte та виводимо текст
kelly = people[people.name == "Kelly Ayotte"]
kelly.iloc[0].text

'kelly ann ayotte et ayot born june 27 1968 is an american politician and the junior united states senator from new hampshire serving since 2011 a member of the republican party she is the second youngest of the 20 female senators and the eleventhyoungest overallborn in nashua new hampshire ayotte is a graduate of pennsylvania state university and villanova university school of law she worked as a law clerk for the new hampshire supreme court before entering private practice she also worked as a prosecutor for the new hampshire department of justice and briefly served as the legal counsel to new hampshire governor craig benson before returning to the department of justice to serve as deputy attorney general of new hampshire in june 2004 governor benson appointed ayotte as attorney general of new hampshire after the resignation of peter heed she became new hampshires first woman attorney general serving from 2004 to 2009 after she was twice reappointed by democratic governor john lynch 

In [164]:
# Знаходимо подібність (відстань) між Керрі та Келлі за косинусною мірою подібності
sim_result = cosine_similarity(dtm[carrie.index], dtm[kelly.index])
print(f"Косинусна подібність між Керрі та Келлі: {sim_result[0,0]:.3f}")

Косинусна подібність між Керрі та Келлі: 0.016


#### Інтерпретація результатів
Отримане значення косинусної подібності 0.016 вказує на дуже низьку схожість між текстами про Керрі та Келлі. Це означає, що ці тексти майже не мають спільних слів або тем.

3. Побудувати NNM-модель на базі TF_IDF (використовуючи міру для відстані
cosine). Знайти найближчих сусідів для людини по варіанту із таблиці завдання Вивести тексти для перших двох найближчих сусідів, перевірити правильність роботи моделі.

In [165]:
# Будуємо NNM-модель
knn_model_1 = NearestNeighbors(n_neighbors=5, metric='cosine')
knn_model_1.fit(dtm)
# Шукаємо найближчих людей до Келлі
distances_1, indices_1 = knn_model_1.kneighbors(dtm[carrie.index], return_distance=True)
# Виводимо таблицю знайдених найближчих людей
carrie_nearest_df = pd.DataFrame({'names': people.loc[indices_1[0]].name, "distance": distances_1[0]})
carrie_nearest_df.sort_values(by="distance").head(5)

names  distance
317    Carrie Underwood  0.000000
29297    Kelly Clarkson  0.632841
508          Faith Hill  0.652212
27793             Adele  0.653539
38592   Billy Ray Cyrus  0.653639

In [166]:
# Виводимо текст для перших двох найближчих сусідів
kelly_clarkson = people[people.name == "Kelly Clarkson"]
kelly_clarkson.iloc[0].text

'kelly brianne clarkson born april 24 1982 is an american singer and songwriter in 2002 she rose to fame after winning the first season of american idol and has since been established as the original american idol her debut single a moment like this topped the us billboard hot 100 and broke the record for the biggest jump to number one in the charts history it became the bestselling single of the year in the country she became the runnerup of world idol the following yearclarksons debut studio album thankful 2003 has been certified 2 platinum and sold over 45 million copies internationally its lead single miss independent became an international hit earning clarkson her first grammy nomination she developed a rockoriented sound with the release of her second album breakaway 2004 it has been certified 6 platinum and sold over 15 million copies worldwide earning clarkson two grammy awards including one for the hit single since u been gone she later took full creative direction of her thi

In [167]:
faith_hill = people[people.name == "Faith Hill"]
faith_hill.iloc[0].text

'faith hill born audrey faith perry september 21 1967 is an american country pop singer and occasional actress she is one of the most successful country artists of all time having sold more than 40 million records worldwide hill is married to country singer tim mcgraw with whom she has recorded several successful duetshills first two albums take me as i am 1993 and it matters to me 1995 established her as a popular country singer they placed a combined three number ones on billboards country charts and were major successes in north america she later rose to mainstream crossover and international fame with the release of her next two albums faith 1998 and breathe 1999 faith spawned her first international hit this kiss and went multiplatinum in various countries breathe became her bestselling album to date and one of the bestselling country albums of all time with the huge crossover success of her signature song breathe it had massive sales worldwide and earned hill three grammy awards 

#### Інтерпретація результатів
Найближчими сусідами є: Kelly Clarkson — відстань ≈ 0.63, Faith Hill — відстань ≈ 0.65.

Це означає, що статті про Kelly Clarkson та Faith Hill найбільш схожі за вмістом на текст про Carrie Underwood з усієї вибірки.

Після виведення змісту текстів, можна зрозуміти, що тексти мають семантичну схожість та переконатися в правильності роботи моделі.